compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -853,7 +853,16 @@ fn transform_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, options: TransformTyOptio
         }
 
         ty::Generator(def_id, substs, movability) => {
-            ty = tcx.mk_generator(*def_id, transform_substs(tcx, substs, options), *movability);
+            // Encode parent substs only
+            ty = tcx.mk_generator(
```

Comment:
This is not a valid generator type -- you probably should be handling the `substs.as_generator().parent_substs()` part in `encode_ty`, not `transform_ty`.

Here, you should just be splitting the generator substs into its parts, transforming the parent substs, and then reconstructing the generator ty from that.

Reply:
Humm, I don't quite understand. The DefId and substs are the only values used in encoding--it doesn't need to be a valid type. (We do the same for other types.) Am I missing something?

The generators also need to match in the substitution dictionary, so transforming this way here ensures they'll be all the same in the dictionary.

Reply:
> We do the same for other types.

That's not great. You should not be constructing types that are invalid. If you really need to collapse types into some minimal representation to use in some dictionary, you shouldn't be using `Ty<'tcx>` at all, to be honest.

Reply:
Sorry, to be clear, you shouldn't be constructing types that are invalid because the compiler has hard-coded assumptions about the structure of these types, and you should be following those. I'm actually gonna probably make these assumptions even more hard coded in #112189.

For example, that CI failure that just occurred (https://github.com/rust-lang/rust/pull/111914#issuecomment-1572877008) is because you're misusing the `ty::Generator` variant, which you should not be doing.

That's a risk because:
1. it introduces more ways for the compiler to crash
2. it makes it harder to review and refactor this code in the future

Among other reasons.

Reply:
Makes sense. Thanks for the explanation! I may still need to create/transform these generators in encode_ty for the substitution dictionary though, but I'll ensure they're are structurally valid. (I'll need some time to think and work on refactoring it if we decide to use something other than `Ty<'tcx>` in the substitution dictionary in the future.)

Reply:
I guess let's compromise here -- you can transform the substs normally, but add a no-op branch for `ty::GeneratorWitness` in `transform_ty`. 

Then in `encode_ty`, you should continue to ICE if you encounter `ty::GeneratorWitness`, but you shouldn't ever hit it because you'll do the parent substs-only encoding as suggested above.

Reply:
Sounds good to me! Will do it.

Reply:
Done. Let me know if it looks good to you!